setup-env

```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+Examples:
+- To create a new Yocto build directory:
+  $ MACHINE=jetson-tx2 DISTRO=testdistro source $PROGNAME build
+- To use an existing Yocto build directory:
+  $ source $PROGNAME build
+"
     fi
+}
+
+clean_up()
+{
+   unset LIST_MACHINES VALID_MACHINE
+   unset CWD TEMPLATES SHORTOPTS LONGOPTS ARGS PROGNAME
+   unset generated_config updated
+   unset MACHINE SDKMACHINE DISTRO OEROOT TEMPLATECONF BBPATH
+}
+
+# get command line options
+SHORTOPTS="h"
+LONGOPTS="help"
+
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
+# Print the usage menu if invalid options are specified
+if [ $? != 0 -o $# -lt 1 ]; then
+   usage && clean_up
+   return 1
+fi
+
+eval set -- "$ARGS"
+while true;
+do
+    case $1 in
+        -h|--help)
+           usage
+           clean_up
+           return 0
+           ;;
+        --)
+           shift
+           break
+           ;;
+    esac
+done
+
+if [ "$(whoami)" = "root" ]; then
+    echo "ERROR: do not use the BSP as root. Exiting..."
+fi
+
+if [ ! -e $1/conf/local.conf ]; then
+    build_dir_setup_enabled="true"
 else
-    mkdir -p "$BDIR"
-    BDIR=`readlink -f "$BDIR"`
+    build_dir_setup_enabled="false"
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$MACHINE" ]; then
+    usage
+    echo -e "ERROR: You must set MACHINE when creating a new build directory."
+    clean_up
+    return 1
 fi
+
+if [ -z "$SDKMACHINE" ]; then
+    SDKMACHINE='x86_64'
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$DISTRO" ]; then
+    usage
+    echo -e "ERROR: You must set DISTRO when creating a new build directory."
+    clean_up
+    return 1
+fi
+
 # Allow for multiple layer sets,
 # default just "layers"
 if [ "x$BMETA" = "x" ]; then
     BMETA="layers"
 fi
-OEROOT=`readlink -f "$OEROOT/$BMETA"`
+OEROOT=`readlink -f "$CWD/$BMETA"`
 export OEROOT
-export BITBAKEDIR="$OEROOT/bitbake"
-. $OEROOT/scripts/oe-buildenv-internal && \
-    $OEROOT/scripts/oe-setup-builddir && \
-    [ -n "$BUILDDIR" ] && cd $BUILDDIR
-TEMPLATECONF=`cat $BUILDDIR/conf/templateconf.cfg`
+
+export TEMPLATECONF=${TEMPLATECONF:-$OEROOT/meta-testdistro/conf}
+
+. $OEROOT/openembedded-core/oe-init-build-env $CWD/$1 > /dev/null
+
+# if conf/local.conf not generated, no need to go further
+if [ ! -e conf/local.conf ]; then
+    clean_up && return 1
+fi
+
+generated_config=
+if [ "$build_dir_setup_enabled" = "true" ]; then
+
+    # Change settings according environment
+    sed -e "s,MACHINE ??=.*,MACHINE ??= '$MACHINE',g" \
+        -e "s,SDKMACHINE ?=.*,SDKMACHINE ?= '$SDKMACHINE',g" \
+        -e "s,DISTRO ?=.*,DISTRO ?= '$DISTRO',g" \
+        -e "s,PACKAGE_CLASSES ?=.*,PACKAGE_CLASSES ?= '$PACKAGE_CLASSES',g" \
+        -i conf/local.conf
+
+    for s in $HOME/.oe $HOME/.yocto; do
+        if [ -e $s/site.conf ]; then
+            echo "Linking $s/site.conf to conf/site.conf"
+            ln -s $s/site.conf conf
+        fi
+    done
```

Comment:
This too looks strange to me. I'm OK with adding something like this (assuming I'm understanding what it's for), but having it be per-distro would be nice.

Reply:
I've seen this done in other distros and I like the pattern. Usually these kinds of settings are applicable across multiple builds -- to use a common download directory or point a few different builds at the same pr server. You'll see later on in the script that if it's not picked up globally, it will get pulled out of templateconf.